b3sum/src/main.rs

```diff
@@ -166,31 +185,46 @@ fn main() -> Result<()> {
         blake3::Hasher::new()
     };
     let print_names = !args.is_present(NO_NAMES_ARG);
+    let raw_output = args.is_present(RAW_ARG);
     let mut did_error = false;
+
     if let Some(files) = args.values_of_os(FILE_ARG) {
-        for filepath in files {
-            let filepath_str = filepath.to_string_lossy();
-            match hash_file(&base_hasher, filepath) {
-                Ok(output) => {
-                    write_output(output, len)?;
-                    if print_names {
-                        println!("  {}", filepath_str);
-                    } else {
-                        println!();
+        if raw_output && files.len() > 1 {
+            did_error = true;
+            eprintln!("b3sum: Only one filename can be provided when using --raw");
+        } else {
+            for filepath in files {
+                let filepath_str = filepath.to_string_lossy();
+                match hash_file(&base_hasher, filepath) {
+                    Ok(output) => {
+                        if raw_output {
+                            write_raw_output(output, len)?;
+                        } else {
+                            write_hex_output(output, len)?;
+                            if print_names {
+                                println!("  {}", filepath_str);
+                            } else {
+                                println!();
+                            }
+                        }
+                    }
+                    Err(e) => {
+                        did_error = true;
+                        eprintln!("b3sum: {}: {}", filepath_str, e);
```

Comment:
I'm a little skeptical of this change. A common case where you hit an error here is when you do something like `b3sum *`, when the current directory also contains subdirectories. You'll see something like this:

```
$ b3sum *
af1349b9f5f9a1a6a0404dea36dcc9499bcb25c9adc112b7cc9a93cae41f3262  a
b3sum: b: Is a directory (os error 21)
af1349b9f5f9a1a6a0404dea36dcc9499bcb25c9adc112b7cc9a93cae41f3262  c
```

I'd expect all that output to stick together if I e.g. redirected to a file with `b3sum * > out.txt`. But with this change, only the successes would go to the file, and the errors would stay in the terminal. That seems kind of weird to me. What do you think?

Reply:
Whichever way we go, we should probably add a comment above this line explaining how we feel about this issue.

Reply:
This is my two-cents opinion.  But I'm happy to do whatever you think is best. 

When we have a command-line tool that might "partially fail / partially succeed", as is the case here, we should write only valid expected output to stdout, and leave all the human-readable error messaging to stderr.  This is because we might be piping the output to some other process that will be parsing the output, and partial failure might be perfectly OK (As long as they are reported). 

We then use exit codes (as we are doing), to signal if there were any errors at all. 

Again - this is my two-cents opinion.  Happy to take your direction on whatever you choose.

Reply:
@oconnor663 I think in the case you've mentioned it makes complete sense for failures to go into the error; if someone wanted the errors to go in standard output they could feed the error output into the same stream. Splitting them up like that is a benefit!

Reply:
I was on the fence about this, so as a tiebreaker I looked at what Coreutils does (`md5sum`, `b2sum`, etc). Looks like they all use stderr for the per-file errors. That's good enough for me :)